lib/install/management/configure.go

```diff
@@ -98,49 +102,150 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
 
 	snapshotName := fmt.Sprintf("%s %s", ConfigurePrefix, conf.Version.BuildNumber)
 	snapshotName = strings.TrimSpace(snapshotName)
-
 	// check for old snapshot
-	// #nosec: Errors unhandled.
-	oldSnapshot, _ := d.appliance.GetCurrentSnapshotTree(d.op)
+	oldSnapshot, err := d.appliance.GetCurrentSnapshotTree(d.op)
+	if err != nil {
+		// log the error but continue
+		d.op.Debugf("Error checking appliance snapshot tree during %s: %s", d.Action.String(), err.Error())
+	}
 
 	newSnapshotRef, err := d.createSnapshot(snapshotName, "configure snapshot")
 	if err != nil {
-		d.deleteUpgradeImages(ds, settings)
+		d.deleteISOs(ds, settings)
+		return err
+	}
+
+	err = d.update(conf, settings)
+	if err != nil {
+		// Roll back
+		d.op.Errorf("Failed to %s: %s", d.Action.String(), err)
+		d.op.Infof("Rolling back %s", d.Action.String())
+
+		if rerr := d.rollback(conf, snapshotName, settings); rerr != nil {
+			d.op.Errorf("Failed to revert appliance to snapshot: %s", rerr)
+			return err
+		}
+		d.op.Infof("Appliance is rolled back to previous version")
+		d.deleteISOs(ds, settings)
+		d.deleteSnapshot(newSnapshotRef, snapshotName, conf.Name)
+
 		return err
 	}
 
-	err = d.update(conf, settings, isConfigureOp)
+	// if we are upgrading evaluate need for inventory upgrade
+	if d.Action == UpgradeAction && d.session.IsVC() {
+		err := d.updateInventory(conf.Name)
+		if err != nil {
+			// if we returned an error then no Inventory work was
+			// started.  If we failed during update we warn the user.
+			d.op.Errorf("An error prevented inventory changes: %s", err)
+		}
+	}
 
 	// If successful try to grant permissions to the ops-user
-	if err == nil && conf.ShouldGrantPerms() {
-		err = opsuser.GrantOpsUserPerms(d.op, d.session, conf)
+	if conf.ShouldGrantPerms() {
+		// TODO: In a DRS environment that's not been upgraded to 1.4
+		// this will attempt to apply permissions to the VmFolder. Does
+		// that work & do we really want that?
+		err := opsuser.GrantOpsUserPerms(d.op, d.session, conf)
 		if err != nil {
-			// Update error message and fall through to roll back
 			err = errors.Errorf("Failed to grant permissions to ops-user, failure: %s", err)
 		}
 	}
 
+	// compatible with old version's upgrade snapshot name
+	if oldSnapshot != nil && (vm.IsConfigureSnapshot(oldSnapshot, ConfigurePrefix) || vm.IsConfigureSnapshot(oldSnapshot, UpgradePrefix)) {
+		d.deleteSnapshot(&oldSnapshot.Snapshot, oldSnapshot.Name, conf.Name)
+	}
+
+	return nil
+}
+
+// updateInventory updates VCH Folder inventories.
+//
+// If the VCH resides in the VMFolder then a new folder will
+// be created and the VCH and all Containers will be moved into the
+// new folder.  If the VCH already resides in a folder other than
+// the VMFolder, then no action will be taken.
+//
+// vApp support planned: https://github.com/vmware/vic/issues/7670
+func (d *Dispatcher) updateInventory(name string) error {
+	defer trace.End(trace.Begin("", d.op))
+
+	appName := d.appliance.Name()
```

Comment:
Should this be `ObjectName()`? I notice feature/robo has changed `vm.Name` to `vm.ObjectName` in most locations.

Reply:
Yeah...we could...this was a quick change this AM on flight to investigate an issue...with this we don't make a property collector as it determines the name off the path..

ultimately `vic-machine` needs to get _most_ if not all the data it needs on init...currently it gets a lot of data multiple times.  So here I used the `.Name()` to avoid yet another round trip...